### PR TITLE
InputControl: Allow onBlur for empty values to commit the change, move reset behaviour to ESCAPE key

### PR DESCRIPTION
**Pull request by andrewserong**
## Description

As reported in #39090, the existing default behaviour of the BoxControl (as used in Padding and Margin dimensions controls) where you need to press Enter to commit a value change such as clearing out the padding value, can be confusing for users when it comes to clearing out the value.

Prior to this PR, when you blur (e.g. tab or click) out of the UnitControl (with `isPressEnterToChange` set to true) when it contains no value, this will be treated as cancelling / resetting the current change.

After this PR, we treat any change to the input field (even it's empty) as a change to commit when tabbing or clicking out of the field.

To preserve the ability to cancel the current change, the `reset()` behaviour is moved to the ESCAPE key, which has some consistency with the [W3C WAI recommendations](https://www.w3.org/TR/wai-aria-practices-1.1/#gridNav_inside) for editing within a data cell (namely `Escape`: if content was being edited, it may also undo edits) — it's not an exact match to the use case of this component, but I think there's a close enough parallel.

The proposed change here is an opinionated one, but hopefully it helps alleviate some of the frustration with editing Padding and Margin values as in the initial report.

## Changes included:

* When `isPressEnterToChange` is set on the InputField, commit empty values
* When `isPressEnterToChange` is set on the InputField, allow the `ESCAPE` key to cancel the current (not-yet-committed) change

## Testing Instructions

* Add a group or row block to a post or in the site editor.
* Adjust the padding and/or margin so that it has a value.
* Save and reload the editor.
* Go to clear out the value by pressing backspace in the field.
* Click outside of the field — the cleared value should persist.
* Save and reload the editor — the cleared value should persist.
* Also check that you're still able to enter a value + unit to switch the unit in the control. E.g. try entering `10px` or `2rem` or `5%` and then press enter and check that it still works correctly.

Confirm existing tests pass (you can look at the CI job, but I ran the following locally to confirm the tests were passing):

```
npm run test-unit packages/components/src/box-control/test
npm run test-unit packages/components/src/input-control/test
npm run test-unit packages/components/src/unit-control/test
```

Also check that the components still behave correctly in Storybook (via `npm run storybook:dev`). Test the BoxControl component in particular: http://localhost:50240/?path=/story/components-experimental-boxcontrol--default

## Screenshots 

| Before | After |
| --- | --- |
| box-control-before | box-control-after |

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

**Comment by andrewserong** [comment]
I thought I'd have a go at this one — it looks like it might not be as simple as it looked superficially — we also need to support being able to type in the unit, and switching this flag to `false` appears to disable that. I'll do some more digging.

**Comment by ramonjd** [comment]
Thanks for working on this and all the copy+paste commands in the test description 😄 

I tested the padding/margin/blockGap inputs (and linked sides where available) in both the block and site editors.

✅  Hitting `ENTER` saves a value in the field
✅  Hitting `ESCAPE` cancels any changes to the field (the original is restored)
✅  Unit value changes are respected (an unaffected)
✅  Tests pass
✅  Storybook looks 👍 
❓  Blurring out of the field works as I'd expect in the site editor and in the post editor, however in the latter, blurring and giving focus to the editor canvas doesn't

**Comment by andrewserong** [comment]
> Blurring out of the field works as I'd expect in the site editor and in the post editor, however in the latter, blurring and giving focus to the editor canvas doesn't

Interesting, good catch @ramonjd! I think that might be because the component is unmounted before `onBlur` has a chance to fire, since in that case the user has clicked somewhere that deselects the block. Not sure the best way to handle that one 🤔

File: packages/components/src/box-control/test/index.js [hunk 1]
```diff
@@ -124,6 +124,24 @@ describe( 'BoxControl', () => {
 			expect( input.value ).toBe( '' );
 			expect( unitSelect.value ).toBe( 'px' );
 		} );
+
+		it( 'should persist cleared value when focus changes', () => {
+			const { container } = render( <BoxControl /> );
+			const input = container.querySelector( 'input' );
+			const unitSelect = container.querySelector( 'select' );
+
```
Comment (ramonjd) [inline packages/components/src/box-control/test/index.js:134]: Would the following also work?

```js
const input = screen.getByLabelText( 'Box Control', {
				selector: 'input',
			} );
const unitSelect = screen.getByLabelText( 'Select unit' );
```

Noting after a dogmatic reading of [React Testing Library's docs](https://testing-library.com/docs/queries/about#manual-queries).

😄

**Comment by ramonjd** [comment]
> I think that might be because the component is unmounted before onBlur has a chance to fire, since in that case the user has clicked somewhere that deselects the block. Not sure the best way to handle that one 🤔

What I forgot to say is that it didn't really stand out to me as an issue. Being able to edit values using the keyboard and via mouse clearing is already a huge win.

File: packages/components/src/box-control/test/index.js (hunk 1 above)
Comment (andrewserong) [reply packages/components/src/box-control/test/index.js:134]: Oh, yes — good to be in keeping with the suggested way of doing it (I was just copying and pasting from other tests 😁). I've updated this one in https://github.com/WordPress/gutenberg/pull/39109/commits/106dfa816dc6ad1755cf26c438ac83086ec84e96, but left the others as-is to keep the size of this PR contained (happy to follow-up on the others separately, though, if we think it's worth it!)

**Review by ciampo — APPROVED** [review]
Excellent work, @andrewserong (and thank you for the ping!) I think this change is an improvement over the previous behaviour 🚀 

**Before**

https://user-images.githubusercontent.com/1083581/156158695-004c6ecb-7341-4dda-8306-06a7fe1eb981.mp4

**After**

https://user-images.githubusercontent.com/1083581/156158707-2b68e521-6306-4a3a-8f9e-77e08b9d4ccc.mp4

---

Regarding the `onBlur` callback not being fired when unmounting the component, I've found a potential workaround (although it's definitely something that can happen a follow-up, since it's not really related to the changes in this PR) highlighted in https://github.com/facebook/react/issues/12363#issuecomment-655441541

**Review by mirka — APPROVED** [review]
Good call, sounds good to me too 👍 

Just curious, any reason we're not adding a test for `InputControl` itself? (`<InputControl isPressEnterToChange />`)

**Comment by stokesman** [comment]
I think the changes here look fine but I am wondering if the option to stop using `isPressEnterToChange={ true }` for these `UnitControl`s should be discussed. It was raised in #30222 but didn't get much input. I'm ambivalent on it but it might be a way optimize the experience for the average user.

**Comment by andrewserong** [comment]
Thanks for the reviews and feedback, folks!

> Just curious, any reason we're not adding a test for `InputControl` itself? (`<InputControl isPressEnterToChange />`)

Good question — I took a look at the existing tests and the tests at the `InputField` level looked fairly minimal when it came to testing keyboard / commit behaviour, I suspected possibly because the `InputField` is mostly just calling passed in functions? It seemed a bit more useful to add a test at more of an integration level in the `UnitControl` and `BoxControl` components, where the behaviour is covered in closer to real world usage, and we already have similar test coverage. Happy to add more tests to `InputField` / `InputControl` in a follow-up if you think we should have more granular tests, though!

> I think the changes here look fine but I am wondering if the option to stop using `isPressEnterToChange={ true }` for these `UnitControl`s should be discussed.

That's a great question, too. I had a go at switching it off before opening this PR, but noticed as you mention in that issue (https://github.com/WordPress/gutenberg/issues/30222#issuecomment-809736646) that it then switched off the behaviour where you can type in the unit and then press Enter to switch units. Without knowing whether or not anyone currently relies on that option, I thought I'd open this PR in the shorter-term to at least improve the UX of this mode. But I think it'd be worth exploring if folks think it's worth it to switch it off on the BoxControl!

File: packages/components/CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
 
 -   `TreeGrid`: Add tests for `onCollapseRow`, `onExpandRow`, and `onFocusRow` callback functions. ([#38942](https://github.com/WordPress/gutenberg/pull/38942)).
 -   `TreeGrid`: Update callback tests to use `TreeGridRow` and `TreeGridCell` sub-components. ([#39002](https://github.com/WordPress/gutenberg/pull/39002)).
+-   InputControl: Allow `onBlur` for empty values to commit the change when `isPressEnterToChange` is true, and move reset behavior to the ESCAPE key. ([#39109](https://github.com/WordPress/gutenberg/pull/39109)).
```
Comment (stokesman) [inline packages/components/CHANGELOG.md:16]: Late comment but this missed the Unreleased section (easy to do when it's empty). Maybe someone can move it in another PR before the release.

Comment (ciampo) [reply packages/components/CHANGELOG.md:16]: Good catch!

Fix in #39169
